rc/tools/clipboard.kak

```diff
@@ -0,0 +1,94 @@
+declare-option -docstring %{
+    The shell command to which data will be piped to copy to the system clipboard
+} str clipboardcmd_copy %sh{
+    for cmd in "xsel -i" "pbcopy" "wl-copy -p"; do
+        readonly program="${cmd%% *}"
+        if command -v "${program}" >/dev/null 2>&1; then
+            printf %s "${cmd}"
+            exit
+        fi
+    done
+}
+
+declare-option -docstring %{
+    The shell command which will paste the system clipboard into the current buffer
+} str clipboardcmd_paste %sh{
+    for cmd in "xsel -o 2>/dev/null" "pbpaste" "wl-paste -p"; do
+        readonly program="${cmd%% *}"
+        if command -v "${program}" >/dev/null 2>&1; then
+            printf %s "${cmd}"
+            exit
+        fi
+    done
+}
+
+define-command -params .. -docstring %{
+    clipboard-copy [<data>…]: copy the given data to the system clipboard
+
+    If no data is passed as argument to the command, the current selection is used
+    If multiple selections/data arguments are to be stored into the clipboard, they are joined with a newline character
+} clipboard-copy %{ evaluate-commands %sh{
+    join() {
+        # NOTE: this is a POSIX way of storing a newline
+        # character into $IFS without spawning a subshell
+        IFS='
+'
+        printf %s "$*"
+    }
+
+    if [ $# -eq 0 ]; then
+        eval "set -- $kak_quoted_selections"
+    fi
+
+    if ! join "$@" | eval "${kak_opt_clipboardcmd_copy}"; then
+        fail "Unable to copy data to the system clipboard"
+    fi
+} }
+
+define-command -docstring %{
+    Paste the contents of the system clipboard before the selection
+} clipboard-paste-before %{
+    execute-keys ! %opt{clipboardcmd_paste} <ret>
+}
+
+define-command -docstring %{
+    Paste the contents of the system clipboard after the selection
+} clipboard-paste-after %{
+    execute-keys <a-!> %opt{clipboardcmd_paste} <ret>
+}
+
+define-command -docstring %{
+    Replace the selection with the contents of the system clipboard
+} clipboard-replace %{
+    evaluate-commands -save-regs c %{
+        set-register c %sh{ eval "${kak_opt_clipboardcmd_paste}" }
+        execute-keys \"c R
+    }
+}
+
+define-command -docstring %{
+    Automatically copy the contents of the `c` register to the clipboard when it is modified
+} clipboard-enable-register %{
+    try clipboard-disable-register
+    hook -group clipboard-bind-register global RegisterModified c %{
+        clipboard-copy %reg{c}
+    }
+}
```

Comment:
Why not take the register to hook as a parameter? I imagine a lot of people would want to `:clipboard-enable-register dquote` — that probably shouldn't be the default, but it should be easy to do.

Reply:
I considered it, but then thought that it'd be better to stay in line with the “pick a good default” line the editor has followed.

I guess that Vi users would rather the register be `+` or `*`, but as the commit message says, punctuation signs are generally used by core primitives.

I also wanted to have the register documented in the docstring so that users wouldn't have to read the code itself, which is difficult to implement if the register is arbitrary.

Reply:
You could also accept a parameter, but default to `c` if there are none.

Reply:
That was the idea, but how to insert the given register into the docstring? It was simpler to hardcode `c` and let users override the command from `autoload`.

Reply:
I'm not sure what is the problem, the docstring can simply mention that the register passed as argument is used, or 'c' if there are none?

Reply:
Well, Kakoune doesn't have `+` or `*` registers, so it doesn't really matter whether punctuation signs are used by core primitives, they're just not available. Besides, Vim also has the `set clipboard += unnamed` option which makes yank/delete/change/put interact with the clipboard register instead of the unnamed register, and I suspect *that's* how a lot of people interact with the clipboard from Vim rather than `"*y`.

Reply:
Came up with this, not sure if this is really better than what we already have:

```
define-command -params ..1 -docstring %{
    clipboard-enable-register [<register>]: automatically copy the contents of the given register to the clipboard, when it is modified

    If no register is passed as argument, the `c` register will be used
    The built-in copy register can be referred to as `dquote`, instead of `"`
} -shell-script-candidates %{
    echo dquote
} clipboard-enable-register %{ evaluate-commands %sh{
    reg="${1:-c}"

    # Make sure the register is valid
    printf 'nop %%reg{%s}\n' "${reg}"

    case "${reg}" in
        dquote) reg='"';;
    esac

    # NOTE: The following will break with the closing-brace register
    printf '
        try clipboard-disable-register
        hook -group clipboard-bind-register global RegisterModified %%{%s} %%{
            clipboard-copy %%reg{%s}
        }
    ' "${reg}" "${reg}"
} }
```

Reply:
I pushed this new implementation. One user indicated that they would like to have seamless integration with the system clipboard when writing prose, but use a letter register for code, which is a good reason to have this more complicated implementation.